### PR TITLE
Add XCM Tracing

### PR DESCRIPTION
**Pull request by apopiak**
This PR adds `log::trace` logging for a selection of XCM methods.
This includes:
+ `log::debug` (should this also be trace?) the generic `execute_xcm` implementation
+ trace the relevant methods in the xcm-executor
+ trace the tuple implementations of various traits in case of failure/filter/miss

The logging choices are tentative and open for feedback. Let me know what should be added/removed.

#### Usage
Add `-lxcm=trace` to any Polkadot node.

### CODE REVIEW

File: xcm/src/v0/traits.rs [hunk 1]
```diff
@@ -134,6 +134,11 @@ pub trait ExecuteXcm<Call> {
 	/// a basic hard-limit and the implementation may place further restrictions or requirements on weight and
 	/// other aspects.
 	fn execute_xcm(origin: MultiLocation, message: Xcm<Call>, weight_limit: Weight) -> Outcome {
+		log::debug!(
+			target: "xcm::execute_xcm",
+			"origin: {:?}, message: {:?}, weight_limit: {:?}",
+			&origin, &message, weight_limit,
+		);
```
Comment (apopiak) [inline xcm/src/v0/traits.rs:141]: log I feel unsure about
Feels useful to be able to just add `-lxcm=debug` and get some measure of logging, but also adds this to (most) every implementation automatically. Also doesn't capture direct calls to `execute_xcm_in_credit` (which e.g. `teleport_assets` does).

**Comment by shawntabrizi** [comment]
Can we confirm this is zero cost to the performance?

Otherwise, a great idea imo

**Comment by apopiak** [comment]
Another general point: Compared to the amount of logging I just added to my Polkadot branch this is quite coarse.

**Comment by apopiak** [comment]
> Can we confirm this is zero cost to the performance? (someone from the core dev team can probably comment)
> 
> Otherwise, a great idea imo

based on my understanding of 
https://github.com/paritytech/substrate/pull/8655/files#diff-58e928d2a8ca73ab1d4a7ef80178314f6144506cba80ff0bd51b8d9928796bf2R44
it should be fine

**Comment by apopiak** [comment]
Basti confirmed that logging is stripped for release builds of wasm runtimes.

**Comment by pepyakin** [comment]
Just a quick and random chime in: if you are interested, there is an easy way to double check. Get the release wasm binary and run `strings wasm.wasm | grep <x>` where `x` is one of the lines you do not expect to be compiled in.

**Review by shawntabrizi — APPROVED** [review]

File: bridges/modules/dispatch/src/lib.rs [hunk 2]
```diff
@@ -142,6 +142,7 @@ impl<T: Config<I>, I: Instance> MessageDispatch<T::MessageId> for Pallet<T, I> {
 		let expected_version = <T as frame_system::Config>::Version::get().spec_version;
 		if message.spec_version != expected_version {
 			log::trace!(
+				target: "runtime::bridge-dispatch",
```
Comment (bkchr) [inline bridges/modules/dispatch/src/lib.rs:145]:
```suggestion
```
Please do this in the bridge repo.

File: xcm/src/v0/traits.rs [hunk 3]
```diff
@@ -134,6 +134,11 @@ pub trait ExecuteXcm<Call> {
 	/// a basic hard-limit and the implementation may place further restrictions or requirements on weight and
 	/// other aspects.
 	fn execute_xcm(origin: MultiLocation, message: Xcm<Call>, weight_limit: Weight) -> Outcome {
+		log::debug!(
+			target: "xcm::execute_xcm",
+			"origin: {:?}, message: {:?}, weight_limit: {:?}",
+			&origin, &message, weight_limit,
```
Comment (bkchr) [inline xcm/src/v0/traits.rs:140]:
```suggestion
			origin,
			message,
			weight_limit,
```

File: xcm/xcm-executor/src/lib.rs [hunk 4]
```diff
@@ -47,6 +47,11 @@ impl<Config: config::Config> ExecuteXcm<Config::Call> for XcmExecutor<Config> {
 		weight_limit: Weight,
 		mut weight_credit: Weight,
 	) -> Outcome {
+		log::trace!(
+			target: "xcm::execute_xcm_in_credit",
+			"origin: {:?}, message: {:?}, weight_limit: {:?}, weight_credit: {:?}",
+			&origin, &message, weight_limit, weight_credit,
```
Comment (bkchr) [inline xcm/xcm-executor/src/lib.rs:53]:
```suggestion
			origin,
			message,
			weight_limit,
			weight_credit,
```

File: xcm/xcm-executor/src/lib.rs [hunk 5]
```diff
@@ -94,6 +100,11 @@ impl<Config: config::Config> XcmExecutor<Config> {
 		maybe_shallow_weight: Option<Weight>,
 		trader: &mut Config::Trader,
 	) -> Result<Weight, XcmError> {
+		log::trace!(
+			target: "xcm::do_execute_xcm",
+			"origin: {:?}, top_level: {:?}, message: {:?}, weight_credit: {:?}, maybe_shallow_weight: {:?}",
+			&origin, top_level, &message, weight_credit, maybe_shallow_weight,
```
Comment (bkchr) [inline xcm/xcm-executor/src/lib.rs:106]:
```suggestion
			origin, 
			top_level, 
			message, 
			weight_credit, 
			maybe_shallow_weight,
```

File: xcm/xcm-executor/src/lib.rs [hunk 6]
```diff
@@ -225,6 +236,11 @@ impl<Config: config::Config> XcmExecutor<Config> {
 		effect: Order<Config::Call>,
 		trader: &mut Config::Trader,
 	) -> Result<Weight, XcmError> {
+		log::trace!(
+			target: "xcm::execute_effects",
+			"origin: {:?}, holding: {:?}, effect: {:?}",
+			origin, holding, &effect,
```
Comment (bkchr) [inline xcm/xcm-executor/src/lib.rs:242]:
```suggestion
			origin,
			holding,
			effect,
```

File: xcm/xcm-executor/src/traits/conversion.rs [hunk 7]
```diff
@@ -173,6 +173,11 @@ impl<O> ConvertOrigin<O> for Tuple {
 				r => return r
 			};
 		)* );
+		log::trace!(
+			target: "xcm::convert_origin",
+			"could not convert: origin: {:?}, kind: {:?}",
+			&origin, &kind
```
Comment (bkchr) [inline xcm/xcm-executor/src/traits/conversion.rs:179]:
```suggestion
			origin,
			kind,
```

File: xcm/xcm-executor/src/traits/filter_asset_location.rs [hunk 8]
```diff
@@ -30,6 +30,11 @@ impl FilterAssetLocation for Tuple {
 		for_tuples!( #(
 			if Tuple::filter_asset_location(what, origin) { return true }
 		)* );
+		log::trace!(
+			target: "xcm::filter_asset_location",
+			"got filtered: what: {:?}, origin: {:?}",
+			&what, &origin,
```
Comment (bkchr) [inline xcm/xcm-executor/src/traits/filter_asset_location.rs:36]:
```suggestion
			what,
			origin,
```

File: xcm/xcm-executor/src/traits/should_execute.rs [hunk 9]
```diff
@@ -58,6 +58,11 @@ impl ShouldExecute for Tuple {
 				_ => (),
 			}
 		)* );
+		log::trace!(
+			target: "xcm::should_execute",
+			"did not pass barrier: origin: {:?}, top_level: {:?}, message: {:?}, shallow_weight: {:?}, weight_credit: {:?}",
+			&origin, top_level, &message, shallow_weight, weight_credit
```
Comment (bkchr) [inline xcm/xcm-executor/src/traits/should_execute.rs:64]:
```suggestion
			origin,
			top_level,
			message,
			shallow_weight,
			weight_credit,
```

File: xcm/xcm-executor/src/traits/transact_asset.rs [hunk 10]
```diff
@@ -107,34 +107,59 @@ impl TransactAsset for Tuple {
 				r => return r,
 			}
 		)* );
+		log::trace!(
+			target: "xcm::TransactAsset::can_check_in",
+			"asset not found: what: {:?}, origin: {:?}",
+			&what, &origin,
```
Comment (bkchr) [inline xcm/xcm-executor/src/traits/transact_asset.rs:113]:
```suggestion
			what,
			origin,
```

File: xcm/xcm-executor/src/traits/transact_asset.rs [hunk 11]
```diff
@@ -107,34 +107,59 @@ impl TransactAsset for Tuple {
...
 	}
+
 	fn deposit_asset(what: &MultiAsset, who: &MultiLocation) -> XcmResult {
 		for_tuples!( #(
 			match Tuple::deposit_asset(what, who) { o @ Ok(_) => return o, _ => () }
 		)* );
+		log::trace!(
+			target: "xcm::TransactAsset::deposit_asset",
+			"did not deposit asset: what: {:?}, who: {:?}",
+			&what, &who,
```
Comment (bkchr) [inline xcm/xcm-executor/src/traits/transact_asset.rs:137]:
```suggestion
			what,
			who,
```

File: xcm/xcm-executor/src/traits/transact_asset.rs [hunk 12]
```diff
@@ -107,34 +107,59 @@ impl TransactAsset for Tuple {
...
 	}
+
 	fn withdraw_asset(what: &MultiAsset, who: &MultiLocation) -> Result<Assets, XcmError> {
 		for_tuples!( #(
 			match Tuple::withdraw_asset(what, who) { o @ Ok(_) => return o, _ => () }
 		)* );
+		log::trace!(
+			target: "xcm::TransactAsset::withdraw_asset",
+			"did not withdraw asset: what: {:?}, who: {:?}",
+			&what, &who,
```
Comment (bkchr) [inline xcm/xcm-executor/src/traits/transact_asset.rs:149]:
```suggestion
			what, 
			who,
```

File: xcm/xcm-executor/src/traits/transact_asset.rs [hunk 13]
```diff
@@ -107,34 +107,59 @@ impl TransactAsset for Tuple {
...
 	}
+
 	fn transfer_asset(what: &MultiAsset, from: &MultiLocation, to: &MultiLocation) -> Result<Assets, XcmError> {
 		for_tuples!( #(
 			match Tuple::transfer_asset(what, from, to) { o @ Ok(_) => return o, _ => () }
 		)* );
+		log::trace!(
+			target: "xcm::TransactAsset::transfer_asset",
+			"did not transfer asset: what: {:?}, from: {:?}, to: {:?}",
+			&what, &from, &to,
```
Comment (bkchr) [inline xcm/xcm-executor/src/traits/transact_asset.rs:161]:
```suggestion
			what,
			from,
			to,
```

**Review by bkchr — APPROVED** [review]
One last thing, using the function name as target sounds like a good idea. However, this is done nowhere else, so yeah, not sure we should start using it here?

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**